File: 240425/갈 수 있는 곳들/places-can-go.cpp
#include <iostream>
#include <queue>

#define MAX_N 100

using namespace std;

queue<pair<int, int>> q;
int grid[MAX_N][MAX_N];
bool vistied[MAX_N][MAX_N];

int n, k;
int dr[4] = {1, -1, 0, 0};
int dc[4] = {0, 0, 1, -1};

bool in_range(int r, int c) {
    return 0 <= r && r < n && 0 <= c && c < n;
}

void bfs() {
    while (!q.empty()) {
        pair<int, int> pos = q.front();
        int r = pos.first, c = pos.second;
        q.pop();

        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i], nc = c + dc[i];
            if (!in_range(nr, nc) || grid[nr][nc] || vistied[nr][nc]) {
                continue;
            }

            q.push(make_pair(nr, nc));
            vistied[nr][nc] = true;
        }
    }
}

int main() {

    cin >> n >> k;
    for (int r = 0; r < n; r++) {
        for (int c = 0; c < n; c++) {
            cin >> grid[r][c];
        }
    }

    for (int i = 0; i < k; i++) {
        int row, col;
        cin >> row >> col;
        q.push(make_pair(row - 1, col - 1));
        vistied[row - 1][col - 1] = true;
    }

    bfs();
    int res = 0;
    for (int r = 0; r < n; r++) {
        for (int c = 0; c < n; c++) {
            if (vistied[r][c]) {
                res++;
            }
        }
    }

    cout << res;
    return 0;
}

File: 240425/갈 수 있는 곳들/places-can-go.py
from collections import deque

n, k = map(int, input().split())

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [
    [False] * n
    for _ in range(n)
]

drs, dcs = [1, 0, -1, 0], [0, 1, 0, -1]

def in_range(r, c):
    return 0 <= r < n and 0 <= c < n

q = deque()
def bfs():
    while q:
        r, c = q.popleft()
        for dr, dc in zip(drs, dcs):
            nr, nc = r + dr, c + dc

            if not in_range(nr, nc) or grid[nr][nc] or visited[nr][nc]:
                continue

            q.append((nr, nc))
            visited[nr][nc] = True

for _ in range(k):
    row, col = map(int, input().split())
    visited[row - 1][col - 1] = True
    q.append((row - 1, col - 1))

bfs()

print(sum([1 for i in range(n) for j in range(n) if visited[i][j]]))